Àlgebra Lineal Numèrica - Curs 2024/2025 - Óscar Rodríguez

---

# Exercicis errors I (Llista T1, ex. 5-8)

## Breu repàs

### Representació en coma flotant


Notem que, genèricament, quan treballem amb la representació de coma flotant, estarem cometent un error. Així, anomenarem $\text{fl}(x)$ a la representació en coma flotant de $x$.

També és usual notar per:
* $\text{fl}_A(x)$ a la representació en coma flotant de $x$ per arrodoniment.
* $\text{fl}_T(x)$ a la representació en coma flotant de $x$ per tall.

&nbsp;

> ###### 🟢 ***Exemple:***
>
> *Sabem que el nombre $\pi$ és $3.1415926535897932384626...$.  Si treballem amb una representació en coma flotant que únicament admet 5 xifres significatives en base 10 aleshores:*
* $\text{fl}_A(\pi) = 0.31416 \cdot 10^1$.
* $\text{fl}_T(\pi) = 0.31415 \cdot 10^1$.  

&nbsp;

### Error absolut i relatiu






Donat un nombre $x$ usualment no som capaços de guardar el seu **valor real** sinó que guardem $\bar{x}$, un **valor aproximat** i, per tant, cometem un error. En particular, analitzarem dos tipus d'errors:

*   **Error absolut**: $e_a(\bar{x}) := x -\bar{x}\,$ (té les mateixes dimensions que $x$).

*   **Error relatiu**: $e_r(\bar{x}) :=\cfrac{x-\bar{x}}{x}\,$  (és adimensional).

**Observació**: Un error absolut d'un metre no és important si estem mesurant la distància entre dos planetes, però pot ser catastròfic en el cas de tenir-lo en la construcció d'una infraestructura.


&nbsp;

En general, no coneixem aquests errors exactament sinó que coneixem un fita d'aquests, $\varepsilon_a(\bar{x})>0$, $\varepsilon_r(\bar{x})>0$ i.e.
$$|e_a(\bar{x})|\leq \varepsilon_a(\bar{x}),  \qquad |e_r(\bar{x})|\leq \varepsilon_r(\bar{x}).$$


**Notació:**

Usualment escriurem:

* $x=\bar{x}+e$ amb $|e|\leq\varepsilon_a(\bar{x})$.

* $\bar{x}=x(1+\delta)$ amb $|\delta|\leq\varepsilon_r(\bar{x})$.
    


&nbsp;



### Estimació dels errors en la representació



Sigui $x\in\mathbb{R}$, $x\neq 0$, tal que $x= \pm 0,d_1d_2\ldots d_n d_{n+1}\ldots \cdot b^e$ amb $d_1\neq 0$. Aleshores:

1) L'error absolut en la representació en coma flotant en base $b$ i $n$ dígits del nombre $x$ satisfà les fites següents:


*   $|e_a(\text{fl}_T(x))| = |\text{fl}_T(x)-x|\leq b^{e-n}\quad \Big(=:\varepsilon_a (\text{fl}_T(x))\Big)$

*   $|e_a(\text{fl}_A(x))| = |\text{fl}_A(x)-x|\leq \cfrac{1}{2}\, b^{e-n}\quad \Big(=:\varepsilon_a (\text{fl}_A(x))\Big)$

&nbsp;

2) L'error relatiu en la representació en coma flotant en base $b$ i $n$ dígits del nombre $x$ satisfà les fites següents:

*   $|e_r(\text{fl}_T(x))| = \left|\cfrac{\text{fl}_T(x)-x}{x}\right|\leq b^{1-n}\quad \Big(=:\varepsilon_r \left(\text{fl}_T(x)\right)\Big)$

*   $|e_r(\text{fl}_A(x))| = \left|\cfrac{\text{fl}_A(x)-x}{x}\right|\leq \cfrac{1}{2}\, b^{1-n}\quad \Big(=:\varepsilon_r (\text{fl}_A(x))\Big)$

&nbsp;

> ###### 🟢 ***Exemple:***
>
>*Tornem a considerar que treballem amb 5 xifres significatives:*
>
> *   $|e_a(\text{fl}_T(\pi))| = |\text{fl}_T(\pi)-\pi| = |0.31415\cdot10^1-\pi|= 0.9265...\cdot10^{-4} \leq 10^{1-5} = 10^{-4}$
>
> *   $|e_a(\text{fl}_A(\pi))| = |\text{fl}_A(\pi)-\pi| = |0.31416\cdot10^1-\pi|= 0.7346...\cdot10^{-5} \leq \cfrac{1}{2}\cdot 10^{1-5} = \cfrac{1}{2}\cdot10^{-4}$
>
> *  $|e_r(\text{fl}_T(\pi))| = \left|\cfrac{\text{fl}_T(\pi)-\pi}{\pi}\right| = \cfrac{|0.31415\cdot10^1-\pi|}{\pi} = 0.2949...\cdot 10^{-4} \leq 10^{1-5} = 10^{-4}$
>
> *  $|e_r(\text{fl}_A(\pi))| = \left|\cfrac{\text{fl}_A(\pi)-\pi}{\pi}\right| = \cfrac{|0.31416\cdot10^1-\pi|}{\pi} = 0.2338...\cdot 10^{-5} \leq \cfrac{1}{2}\cdot10^{1-5} = \cfrac{1}{2}\cdot10^{-4}$

### Propagació de l'error



Siguin $\bar{x}$ i $\bar{y}$ les representacions, afectades per error, de $x$ i $y$, dos nombres reals. Per a les operacions aritmètiques bàsiques $(+,-,\cdot,/)$ tenim les següents fites per l'error absolut:

$
\begin{aligned}
\,\,(1)\,\, & \varepsilon_a(\bar{x}\pm\bar{y}) = \varepsilon_a(\bar{x})+ \varepsilon_a(\bar{y})\\[1.75ex]
(2)\,\, & \varepsilon_a(\bar{x}\cdot\bar{y}) \approx |\bar{y}| \varepsilon_a(\bar{x})+ |\bar{x}|\varepsilon_a(\bar{y})\\
(3)\,\, & \varepsilon_a(\bar{x}/\bar{y}) \approx \cfrac{1}{|\bar{y}|} \varepsilon_a(\bar{x})+ \left|\cfrac{\bar{x}}{\bar{y}^2}\right|\varepsilon_a(\bar{y})\\
\end{aligned}
$

i les següents fites per l'error relatiu:

$
\begin{aligned}
\,\,(4)\,\, & \varepsilon_r(\bar{x}\pm\bar{y}) \approx \left|\cfrac{\bar{x}}{\bar{x}\pm\bar{y}}\right|\varepsilon_r(\bar{x})+ \left|\cfrac{\bar{y}}{\bar{x}\pm\bar{y}}\right|\varepsilon_r(\bar{y})\\
(5)\,\, & \varepsilon_r(\bar{x}\cdot\bar{y}) \approx \varepsilon_r(\bar{x})+ \varepsilon_r(\bar{y})\\[1.75ex]
(6)\,\, & \varepsilon_r(\bar{x}/\bar{y}) \approx \varepsilon_r(\bar{x})+ \varepsilon_r(\bar{y})\\[1.75ex]
\end{aligned}
$


&nbsp;

> ⚠️ **Atenció!**
>
> **Operacions sensibles:** Cal evitar la resta de nombres molt propers (o suma de nombres molt propers en valor absolut i signe diferent).

&nbsp;


#### Fórmules de propagació:

Per problemes numèrics més genèrics on calculem $y=f(x)$ a partir d'una única dada $x$ obtenim la **fórmula aproximada de propagació de l'error**

$$ e_a(y) \approx f'(x)\cdot e_a(x),$$

la qual és conseqüència del [Teorema del valor mitjà](https://ca.wikipedia.org/wiki/Teorema_del_valor_mitj%C3%A0). D'igual forma, tenim la **fórmula aproximada de propagació de l'error maximal**

$$ \varepsilon_a(y) \approx |f'(x)|\cdot \varepsilon_a(x).$$



&nbsp;



### Propagació dels errors als càlculs:

Partint de les dades inicials exactes i, degut a les acumulacions del errors en les operacions, obtenim un resultat amb error. De cara a analitzar aquest error acumulat ho podem fer tenint en compte que l'error es propaga en les operacions aritmètiques bàsiques de la següent manera:

$$\text{fl} (x\square y) = (x\square y) (1+ \delta_\square)$$

on $|\delta_\square| \leq \varepsilon_\square$.

## Precisió en python

Per treballar amb diferents tipus de precisió utilitzarem (depenent del cas) numpy, que ens ofereix els formats estàndard:

| Nom                | Bits | Dígits | Dígits decimals | Bits exponent | Exponent decimal màxim | a numpy                               |
|--------------------|------|--------|-----------------|---------------|-----------------------|------------------------------------------|
| Mitja precisió     | 16   | 11     | 3.31            | 5             | 4.51                  | np.half, np.float16                      |
| Precisió simple    | 32   | 23     | 7.22            | 8             | 38.23                 | np.single, np.float32                    |
| Precisió doble     | 64   | 53     | 15.95           | 11            | 307.95                | np.double, np.float64, np.float_         |
| Precisió quadruple | 128  | 113    | 34.02           | 15            | 4931.77               | (*)  |
| Precisió òctuple   | 256  | 237    | 71.34           | 19            | 78913.20              |                                          |


(*) A numpy també tenim els tipus longdouble, float128 i longfloat que estan fets perquè siguin compatibles amb el long double de C. Aquests, però, no segueixen les regles de l'IEEE 754 i hi ha processadors que no el suporten bé. Depenent l'arquitectura el float128 de numpy pot ser equivalent a __float128, però no és estàndard.







> ⚠️ **Atenció!**
>
>  Augmentar la precisió de les variables incrementa el cost computacional, alentint (notablement) els càlculs.

Per tant, és important triar el tipus de les variables amb les quals volem treballar.

En casos que vulguem treballar amb més precisió, una bona idea és utilitzar alguna llibreria de precisió múltiple, com la llibreria [mpmath](https://mpmath.org/) utilitzant [gmpy2](https://pypi.org/project/gmpy2/). També podem treballar amb precisió exacta o definir un tipus de precisió amb la llibreria decimal. Algunes comandes importants (que farem servir a l'exercici 8) són:

In [ ]:
import decimal as dc

dc.getcontext().prec = 3 # Considerarem els nombres amb 3 xifres significatives
dc.getcontext().rounding = dc.ROUND_DOWN # Imposem el tipus d'arrodoniment, les opcions es poden veure a la taula següent:

| Tipus	             | Estratègia d'Arrodoniment             |
|--------------------|---------------------------------------|
| dc.ROUND_CEILING	 | Arrodoniment cap amunt                |
| dc.ROUND_FLOOR	   | Arrodoniment cap avall                |
| dc.ROUND_DOWN	     | Truncament                            |
| dc.ROUND_UP	       | Arrodoniment lluny de zero            |
| dc.ROUND_HALF_UP	 | Arrodoniment a la meitat cap amunt    |
| dc.ROUND_HALF_DOWN | Arrodoniment a la meitat cap avall    |
| dc.ROUND_HALF_EVEN | Arrodoniment a la meitat cap a parell |
| dc.ROUND_05UP	     | Arrodoniment cap amunt i cap a zero   |

> ⚠️ **Atenció!**
>
>  Tot i que les variables puguin emmagatzemar valors fins a una certa precisió no vol dir que el resultat tingui aquesta precisió!

> ###### 🟢 ***Exemple:***
>
> *Comparem el resultat la precissió que obtenim al guardar una aproximació del nombre $\pi$ en una variable en precissió simple i una altra resultant de sumar 10000 cops aquesta i dividir el resultat per 10000.*

In [ ]:
# Importem la llibreria numpy amb l'àlies np:
import numpy as np


# Imprimim el resultat de pi guardat en numpy (precissió doble per defecte) tot
# mostrant 10 decimals:
print(f"valor pi referència:        {np.pi:.10f}")

# Guardem el valor pi de numpy en un float32:
pi32 = np.float32(np.pi)
print(f"pi en precisió simple:      {pi32:.10f} (error {np.abs(np.pi-pi32):.5e})")

# Fem la operació de sumar 10000 cops i després dividir per 10000:
suma = pi32
for i in range(9999):
    suma +=pi32
suma/=10000
print(f"pi resultant de l'operació: {suma:.10f} (error {np.abs(np.pi-suma):.5e})")

valor pi referència:        3.1415926536
pi en precisió simple:      3.1415927410 (error 8.74228e-08)
pi resultant de l'operació: 3.1411341797 (error 4.58474e-04)


## Exercicis:

### Exercici 5

*Calculeu $f(x) = 1 - \cos(x)$ amb aritmètica de coma flotant de $6$ dígits, per a $|x| < 10^{-3}$. És fiable el resultat? Feu el mateix amb la representació de $f(x)$*
$$
	f(x) = \frac{\sin^{2}(x)}{1 + \cos(x)}.
$$
*Trobeu una altra representació alternativa per a $f(x)$ que sigui vàlida per a $|x|$ petit.*

&nbsp; &nbsp; &nbsp; *(a mà)*

### Exercici 7

Si usem un ordinador que comet errors relatius fitats per $\epsilon$ en la
representació i en les operacions aritmètiques, fiteu l'error comès en el càlcul
de $\sum_{i=1}^{n} x_{i}$.

&nbsp; &nbsp; &nbsp; *(a mà)*

&nbsp;


### Exercici 6

*Se sap que la sèrie $\sum_{n=1}^{\infty}1/n$ és divergent. No obstant això, si la intentem "sumar" en un ordinador usant precisió simple (per qüestions de temps) dóna un valor concret. Trobeu aquest valor i expliqueu aquest fenomen.*

&nbsp;

#### Resolució



Si definim
$\displaystyle{S_n = \sum_{i=1}^n \cfrac{1}{i}},$
volem trobar el valor de $n$ tal que $\text{fl}(S_n) =\text{fl}(S_{n+1})$ utilitzant precisió simple.

Per fer-ho, escriurem un programa en python utilitzant la llibreria numpy.

In [ ]:
# Utilitzarem la llibreria numpy:
import numpy as np

# Inicalitzem les variables
sum_ant = np.float32(0) # precisió simple
sum_new = np.float32(1) # precisió simple
n = 1


# Anirem sumant fins que el nombre no canvii
while sum_new != sum_ant:
    n += 1
    sum_ant = sum_new
    sum_new += np.float32(1/n)

# El nombre n és l'anterior:
n -= 1

# Imprimim el resultat
print(f"Ha sumat els {n} primers nombres, i el seu valor és {sum_new:.8f}") # Afegeixo un format particular per l'output amb 8 xifres decimals

Ha sumat els 2097151 primers nombres, i el seu valor és 15.40368271


El que succeeix és que, a mesura que anem sumant, el terme que afegim és més
petit, de manera que per a $n$ suficientment gran, $n= 2097151$ pel cas de precisió simple, el terme $\cfrac{1}{n+1}$ no afegeix dígits significatius a la suma.

Observem que en aquest cas, el valor que obtenim per $S_n$, és $15.40368271$.

&nbsp;

Observem que si fem la suma en l'ordre invers el resultat que obtenim és diferent:

In [ ]:
# Fem el mateix sumant en l'odre invers els n elements:
sum_cont = np.float32(0)
for k in range(n,0,-1):
  sum_cont += np.float32(1/k)

# Imprimim els resultats de la suma en l'ordre invers:
print(f"La suma en l'odre invers dels {n} primer termes és {sum_cont:.8f}")

La suma en l'odre invers dels 2097151 primer termes és 15.13289833


Amb aquest procediment el resultat calculat per $S_n$  és diferent a l'anterior, ja que ens dona $15.13289833$.

&nbsp;

Aquest segon procediment és més precís que l'anterior (pel que hem vist a l'exercici 7). En efecte, si calculem el mateix en precisió doble i mirem els errors fets, tenim:

In [ ]:
# Repetim el darrer càlcul utilitzant precisió doble:
sum_dob = np.double(0)
for k in range(n,0,-1):
  sum_dob += np.double(1/k)

# Printem el resultat:
print(f"La suma utilitzant precisió doble dels {n} primer termes és {sum_dob:.8f}\n")


# Mirem quin error hem fet comparnt la diferència amb el resultat obtingut amb doble precisió:
dif1 = abs(sum_dob - sum_new)
dif2 = abs(sum_dob - sum_cont)
print(f"L'error obtingut en el primer cas és {dif1:.5e}")
print(f"L'error obtingut en el segon cas és {dif2:.5e}")

La suma utilitzant precisió doble dels 2097151 primer termes és 15.13330622

L'error obtingut en el primer cas és 2.70376e-01
L'error obtingut en el segon cas és 4.07888e-04


### Exercici 8

*Useu aritmètica de $3$ dígits amb tall per a calcular la suma $\displaystyle\sum_{i=1}^{15} \frac{1}{i^{2}}$ primer en l'ordre natural,
$1 + \cfrac{1}{4} + \ldots + \cfrac{1}{225}$, i després a l'inrevés, $\cfrac{1}{225} + \cfrac{1}{196} +\ldots + \cfrac{1}{4} + 1$. Decidiu quin és el mètode més exacte de tots dos.*

#### Resolució

Anomenem $S_{15}$ a la suma en l'ordre natural i $\bar{S}_{15}$ a la suma en l'ordre invers. Per calcular-ho utilitzarem la llibreria decimals.



In [1]:
import decimal as dc
import numpy as np

dc.getcontext().prec = 3 # Considerarem els nombres amb tres xifres significatives
dc.getcontext().rounding = dc.ROUND_DOWN # Tallem els nombres a la tercera xifra significativa (sense aquesta línia fa un arrodoniment)

# Nombre de termes que sumarem:
n = 15


### Ordre natural:
sum_nat = dc.Decimal("0")

for i in range(1,n+1):
    i2 = i*i
    sum_nat += dc.Decimal(1)/dc.Decimal(i2)

print(f"La suma en l'ordre natural {sum_nat:.3e}")


### Ordre invers:
sum_inv = dc.Decimal(0)

for i in range(n,0,-1):
    i2 =i*i
    sum_inv += dc.Decimal(1)/dc.Decimal(i2)

print(f"La suma en l'ordre invers val {sum_inv:.3e}")


### Resultat treballant en doble precisió:
sum_dob = np.double(0)

for i in range(n,0,-1):
  sum_dob += np.double(1/(i*i))

# Printem el resultat:
print(f"La suma utilitzant precisió doble és {sum_dob:.8f}\n")

## Printem els errors:
print(f"Error suma directa: {np.abs(sum_dob-np.double(sum_nat)):.5e}")
print(f"Error suma inversa: {np.abs(sum_dob-np.double(sum_inv)):.5e}")

La suma en l'ordre natural 1.530e+0
La suma en l'ordre invers val 1.570e+0
La suma utilitzant precisió doble és 1.58044028

Error suma directa: 5.04403e-02
Error suma inversa: 1.04403e-02


Pel mateix argument que al problema anterior, la suma en l'ordre invers té una precisió més elevada. En aquest cas, com que treballem amb 3 digits amb tall, prenem $\varepsilon = 10^{1-3} = 10^{-2}$ i podem fitar l'error:

Suma en ordre normal:

$
\left|S_{15} - \left(1 + \cfrac{1}{4} + \ldots + \cfrac{1}{225}\right)\right|
\leq \left(15\cdot 1 + 15\cdot\cfrac{1}{4} + 14\cdot\cfrac{1}{9} + \ldots + 3\cdot\cfrac{1}{196} 	+ 2\cdot\cfrac{1}{225}\right) \cdot 10^{-2} < 0.2255.
$

&nbsp;
Suma en ordre invers:

$
\left|\bar{S}_{15} - \left(\cfrac{1}{225} + \ldots + \cfrac{1}{4} + 1\right)\right|
\leq \left(15\cdot \cfrac{1}{225} + 15\cdot\cfrac{1}{196} + 14\cdot\cfrac{1}{169} + \ldots + 3\cdot\cfrac{1}{4} 	+ 2\cdot1\right) \cdot 10^{-2} < 0.04895.
$
